lib/authentication_with_phone_number/infrastructure/phone_number_authentication_repository.dart

```diff
@@ -0,0 +1,108 @@
+import 'dart:io';
+
+import 'package:dartz/dartz.dart';
+import 'package:firebase_auth/firebase_auth.dart' hide User;
+import 'package:flutter/material.dart';
+import 'package:flutter/services.dart';
+import 'package:fpb/authentication_with_phone_number/domain/i_phone_number_repository_facade.dart';
+import 'package:fpb/core/failures/auth_failure.dart';
+import 'package:injectable/injectable.dart';
+
+@LazySingleton(as: IPhoneNumberRepositoryFacade)
+class PhoneNumberAuthenticationRepository implements IPhoneNumberRepositoryFacade {
+
+  final FirebaseAuth _firebaseAuth;
+
+  PhoneNumberAuthenticationRepository(this._firebaseAuth);
+
+
+  @override
+  Future<Either<AuthFailure, bool>> verifyAndLogin({required String verificationId, required String smsCode}) async {
+    try {
+      final AuthCredential authCredential =
+      PhoneAuthProvider.credential(
+          verificationId: verificationId, smsCode: smsCode);
+
+      final  authResult = await _firebaseAuth.signInWithCredential(authCredential);
+
+      return right(authResult.additionalUserInfo!.isNewUser);
+
+    } on PlatformException catch (e) {
+      debugPrint("Error occurred: code: ${e.code} , message: ${e.message}");
+      if (e.code == 'ERROR_INVALID_CREDENTIAL') {
+        return left(const AuthFailure.invalidCredential());
+      } else if (e.code == 'ERROR_USER_DISABLED') {
+        return left(const AuthFailure.userDisable());
+      } else if (e.code == 'ERROR_ACCOUNT_EXISTS_WITH_DIFFERENT_CREDENTIAL') {
+        return left(const AuthFailure.accountExitWithDifferentCred());
+      } else if (e.code == 'ERROR_OPERATION_NOT_ALLOWED') {
+        return left(const AuthFailure.operationNotAllowed());
+      } else if (e.code == 'ERROR_INVALID_VERIFICATION_CODE') {
+        return left(const AuthFailure.invalidVerificationCode());
+      } else if (e.code == 'ERROR_INVALID_ACTION_CODE') {
+        return left(const AuthFailure.invalidVerificationCode());
+      } else {
+        debugPrint("Error occurred: code: ${e.code} , message: ${e.message}");
+        return left(const AuthFailure.serverError());
+      }
+    }
+  }
+
+
+  Future<Either<AuthFailure, Unit>> verifyOTPCode({
+    required String phoneNumber,
+    required Duration timeOut,
+    required PhoneVerificationFailed phoneVerificationFailed,
+    required PhoneVerificationCompleted phoneVerificationCompleted,
+    required PhoneCodeSent phoneCodeSent,
+    required PhoneCodeAutoRetrievalTimeout autoRetrievalTimeout,
+  }) async {
+    try {
+      _firebaseAuth.verifyPhoneNumber(
```

Comment:
What you can do here is define handlers for ```PhoneVerificationFailed```, ```PhoneVerificationCompleted```, ```PhoneCodeSent```, ```PhoneCodeAutoRetrievalTimeout``` as accessory and private functions inside this Repository class and then call them here. If you realize, you see that ```verificationId``` is set and used only in this repository, so one thing you can do here is set is as a class variable initialized with an empty string, set the value in ```PhoneCodeSent``` and then use it inside ```verifyAndLogin```